src/Acuminator/Acuminator.Analyzers/StaticAnalysis/DacPropertyAttributes/DacPropertyAttributesAnalyzer.cs

```diff
@@ -44,66 +48,170 @@ private static Task AnalyzePropertyAsync(SymbolAnalysisContext symbolContext, PX
 		}
 
 		private static async Task CheckDacPropertyAsync(IPropertySymbol property, SymbolAnalysisContext symbolContext, PXContext pxContext,
-														FieldAttributesRegister fieldAttributesRegister)
+														FieldTypeAttributesRegister fieldAttributesRegister)
 		{
 			ImmutableArray<AttributeData> attributes = property.GetAttributes();
 
 			if (attributes.Length == 0 || symbolContext.CancellationToken.IsCancellationRequested)
 				return;
 
-			var attributesWithInfo = GetFieldAttributesInfos(pxContext, attributes, fieldAttributesRegister, symbolContext.CancellationToken);
+			var attributesWithInfos = GetFieldTypeAttributesInfos(pxContext, attributes, fieldAttributesRegister, symbolContext.CancellationToken);
 
-			if (symbolContext.CancellationToken.IsCancellationRequested || attributesWithInfo.IsNullOrEmpty())
+			if (symbolContext.CancellationToken.IsCancellationRequested || attributesWithInfos.Count == 0)
 				return;
 
-			if (attributesWithInfo.Count > 1)
-			{
-				var locationTasks = attributesWithInfo.Select(info => GetAttributeLocationAsync(info.Attribute, symbolContext.CancellationToken));
-				Location[] attributeLocations = await Task.WhenAll(locationTasks)
-														  .ConfigureAwait(false);
+			bool validSpecialTypes = await CheckForMultipleSpecialAttributesAsync(symbolContext, attributesWithInfos).ConfigureAwait(false);
 
-				foreach (Location attrLocation in attributeLocations)
-				{
-					if (attrLocation == null)
-						continue;
+			if (!validSpecialTypes || symbolContext.CancellationToken.IsCancellationRequested)
+				return;
 
-					symbolContext.ReportDiagnostic(Diagnostic.Create(Descriptors.PX1023_DacPropertyMultipleFieldAttributes, attrLocation));
-				}
-			}
-			else
-			{
-				var (fieldAttribute, fieldAttrInfo) = attributesWithInfo[0];
-				await CheckAttributeAndPropertyTypesForCompatibilityAsync(property, fieldAttribute, fieldAttrInfo, pxContext, symbolContext);
-			}
+			await CheckForFieldTypeAttributesAsync(property, symbolContext, pxContext, attributesWithInfos)
+					.ConfigureAwait(false);
 		}
 
-		private static List<(AttributeData Attribute, FieldAttributeInfo Info)> GetFieldAttributesInfos(PXContext pxContext,
+		private static List<(AttributeData Attribute, List<FieldTypeAttributeInfo> Infos)> GetFieldTypeAttributesInfos(PXContext pxContext,
```

Comment:
@dropsonic I don't think we should use immutable collections everywhere - this is private method and part of implementation, it is not stored anywhere or passed outside the class.
 I checked the Roslyn sources - indeed they use Immutable collections instead of common ones but they also use mutable collections in the private methods - something like ArrayBuilder

Reply:
@SENya1990 ok